common/grammar-parser.cpp

```diff
@@ -137,6 +152,87 @@ namespace grammar_parser {
             bool                                 is_nested) {
         size_t last_sym_start = out_elements.size();
         const char * pos = src;
+
+        auto handle_repetitions = [&](int min_times, int max_times) {
+
+            if (last_sym_start == out_elements.size()) {
+                throw std::runtime_error(std::string("expecting preceding item to */+/?/{ at ") + pos);
+            }
+
+            // apply transformation to previous symbol (last_sym_start to end) according to
+            // the following rewrite rules:
+            // S*     --> S{0,}
+            // S+     --> S{1,}
+            // S?     --> S{0,1}
+            // S{m,n} --> S'     ::= Scopy Scopy Scopy... (m times) S(n-m)
+            //            Scopy  ::= S
+            //            S(x)   ::= Scopy S(x-1) |
+            //            S(x-1) ::= Scopy S(x-2) |
+            //            S(1)   ::= Scopy |
+            // S{m,} -->  S'     ::= Scopy Scopy Scopy (m times) Sstar
+            //            Scopy  ::= S
+            //            Sstar  ::= Scopy Sstar |
+            // And if S is a reference to a rule, then we skip the S_copy indirection
```

Comment:
```suggestion
            // And if S is a reference to a rule, then we skip the Scopy indirection
```

Reply:
I'm unsure if it would be best to have all of these as "S_copy" or "Scopy" in the comments

Reply:
Ah, fair point, will update, I must say it does twist my mind a bit too. I even started w/ `Ssub` and named it copy to avoid that Sslytherin ssound.